src/react/useSpace.ts

```diff
@@ -0,0 +1,48 @@
+import { useContext, useEffect, useRef } from 'react';
+import { SpaceContext } from './contexts/SpaceContext.js';
+import { useChannelState } from './useChannelState.js';
+import { useConnectionState } from './useConnectionState.js';
+
+import type { ErrorInfo } from 'ably';
+import type { Space } from '..';
+import type { UseSpaceCallback, UseSpaceOptions } from './types.js';
+
+interface UseSpaceResult {
+  space?: Space;
+  enter?: Space['enter'];
+  leave?: Space['leave'];
+  updateProfileData?: Space['updateProfileData'];
+  connectionError: ErrorInfo | null;
+  channelError: ErrorInfo | null;
+}
+
+export const useSpace = (callback?: UseSpaceCallback, options?: UseSpaceOptions): UseSpaceResult => {
+  const space = useContext(SpaceContext);
+  const callbackRef = useRef(callback);
+
+  const channelError = useChannelState(space?.channel);
+  const connectionError = useConnectionState(space?.client.connection);
+
+  useEffect(() => {
+    callbackRef.current = callback;
+  }, [callback]);
```

Comment:
isn't using `useRef` with `useEffect` here the same as:

```
const callbackRef = useCallback((...params) => callback(...params), [callback]);
```

Reply:
not, it's a little bit different, `useRef` + `useEffect` will create single container which contains actual `callback` inside, this container won't change after re-render, `useCallback` will create new instance each time `callback` has changed (and usually this callbacks anonymous functions, so it will be on every re-render). If you don't like this hack (which is I believe very common pattern in react libs) we can `subscribe` / `unsubscribe` on every `callback` change. I think `subscribe` / `unsubscribe` is very cheap in our case

Reply:
I see what you mean, this I guess prevents rebinding the useEffect just below during a re-render (because a changing `callback` reference will cause a re-render in both cases right?)

I don't mind the solution, just trying to understand it's benefits better here.

Reply:
Yeah, that's right